src/main/scala/chisel3/PolymorphicIOModule.scala

```diff
@@ -0,0 +1,36 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package chisel3
+
+import chisel3.experimental.{BaseModule, ExtModule, FlatIO, Param}
+
+/** A module or external module whose IO is generated from a specific generator.
+  * This module may have no additional IO created other than what is specified
+  * by its `ioGenerator` abstract member.
+  */
+sealed trait FixedIOBaseModule[A <: Data] extends BaseModule {
+
+  /** A generator of IO */
+  protected def ioGenerator: A
+
+  final val io = FlatIO(ioGenerator)
```

Comment:
couldn't the order of trait-mixing-in allow a user to get around this?

```
trait MySneakyIO {
  val sneakyIO = IO(UInt(32.W))
}

class MyThing extends MySneakyIO with FixedIORawModule 
```

Reply:
Fortunately, no because `FixedIORawModule`/`FixedIOExtModule` are classes and not traits. Those have to be the first thing you mix-in, i.e., they need to follow `extends`.

A user will see the following compile-time error if they try the above:

```
[error] /repos/github.com/chipsalliance/chisel/src/test/scala/chiselTests/FixedIOModuleSpec.scala:55:43: class FixedIORawModule needs to be a trait to be mixed in
[error]     class MyThing extends MySneakyIO with FixedIORawModule[Bool](Bool())
[error]                                           ^
```

If they change the order to be legal Scala, they'll get a runtime error because the trait constructor body runs after IO creation is disallowed:

```
[info]   java.lang.IllegalArgumentException: requirement failed: This module cannot have user-created IO
[info]   at ... ()
[info]   at chiselTests.FixedIOModuleSpec$MySneakyIO$1.$anonfun$sneakyIO$2(FixedIOModuleSpec.scala:52)
[info]   at chisel3.experimental.prefix$.apply(prefix.scala:50)
[info]   at chiselTests.FixedIOModuleSpec$MySneakyIO$1.$anonfun$sneakyIO$1(FixedIOModuleSpec.scala:52)
[info]   at chisel3.internal.plugin.package$.autoNameRecursively(package.scala:33)
[info]   at chiselTests.FixedIOModuleSpec$MySneakyIO$1.$init$(FixedIOModuleSpec.scala:52)
[info]   at chiselTests.FixedIOModuleSpec$MyThing$1.<init>(FixedIOModuleSpec.scala:55)
[info]   at chiselTests.FixedIOModuleSpec.$anonfun$new$9(FixedIOModuleSpec.scala:57)
[info]   at ... ()
[info]   at ... (Stack trace trimmed to user code only. Rerun with --full-stacktrace to see the full stack trace)
```